### PR TITLE
change softmax_lse correction of CP to FP32

### PR DESCRIPTION
**Pull request by xrennvidia**
# Description

- softmax_lse correction is in FP64 now, we can lower it to FP32.
- use `log1p` to be consistent with [PR1401](https://github.com/NVIDIA/TransformerEngine/pull/1401).

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by mgoldfarb-nvidia — APPROVED** [review]
LGTM

**Comment by cyanguwa** [comment]
CI failures are not related to this PR.

`log1p(x)` provides more accuracy when `x` is close to 0.

Downcasting `softmax_lse` from `double` to `float` because cuDNN output is in FP32 and combination of multiple results from different CP ranks should still stay in the `float` limits (see [here](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py#L145)).

**Review by cyanguwa — APPROVED** [review]
